upstream: introduce PriorityStateManager, refactor EDS

dio opened the pull request:
*Description*:
This patch introduces `PriorityStateManager` to manage priorities/hosts/localities
as requirement prior updating the cluster's `PrioritySet`. 
Since most of the codes are taken from eds.cc, this patch also refactors EDS codes.

This is part of the effort on adding `load_assignment` field to `Cluster` as signalled 
in https://github.com/envoyproxy/envoy/pull/3261.

*Risk Level*: Medium
*Testing*: `bazel test //test/...`
*Docs Changes*: N/A
*Release Notes*: N/A

Signed-off-by: Dhi Aurrahman <dio@rockybars.com>

alyssawilk commented on source/common/upstream/eds.cc:
```diff
@@ -63,33 +59,28 @@ void EdsClusterImpl::onConfigUpdate(const ResourceVector& resources, const std::
   }
+
+  auto& priority_state = priority_state_manager_->priority_state_;
+  priority_state.clear();
```
Isn't clearing and creating every time going to be more computationally expensive?  It'd be good if we can do this refactor with minimal performance regression

zuercher replied:
Does it makes sense to make PriorityStateManager a local variable in onConfigUpdate? It would mean reverting the `Config::Utility::checkLocalInfo` call, but other than that check, PSM isn't needed beyond the scope on onConfigUpdate.

alyssawilk commented on source/common/upstream/upstream_impl.h:
```diff
@@ -494,6 +493,56 @@ class ClusterImplBase : public Cluster, protected Logger::Loggable<Logger::Id::u
+  // Returns the size of the current cluster priority state.
+  size_t size() const { return priority_state_.size(); }
+
+  PriorityState priority_state_;
```
I believe per our == google style  this should be private

alyssawilk commented on source/common/upstream/upstream_impl.h:
```diff
@@ -494,6 +493,56 @@ class ClusterImplBase : public Cluster, protected Logger::Loggable<Logger::Id::u
+
+/**
+ * Manages PriorityState of a cluster. PriorityState is a per-priority binding of a set of hosts
+ * with its corresponding locality weight map. This is not used at runtime, but useful to store
```
Sorry, can you clarify what you mean by this not being used at runtime?

dio replied:
Probably the wording is not right. I meant to say: it is configuration related only, not in data path.

alyssawilk commented on source/common/upstream/upstream_impl.cc:
```diff
@@ -645,6 +645,128 @@ ClusterInfoImpl::ResourceManagers::load(const envoy::api::v2::Cluster& config,
+  if (locality_lb_endpoint.has_locality() && locality_lb_endpoint.has_load_balancing_weight()) {
+    // TODO(dio): Add tests as per @alyssawilk's comment here
+    // https://github.com/envoyproxy/envoy/pull/3261/files#r188292098
+    priority_state_[priority].second[locality_lb_endpoint.locality()] =
```
Interestingly this looks like it's covered, but there's a few other spots missing.  It'd be great to have unit tests as part of this refactor.

dio replied:
:+1:

dio commented:
@alyssawilk when you have time please do a more in-depth review 😄. Thank you!

alyssawilk commented on source/common/upstream/eds.cc:
```diff
@@ -63,33 +59,28 @@ void EdsClusterImpl::onConfigUpdate(const ResourceVector& resources, const std::
+                  health_status == envoy::api::v2::core::HealthStatus::DRAINING ||
+                  health_status == envoy::api::v2::core::HealthStatus::TIMEOUT
+              ? absl::optional<Host::HealthFlag>(Host::HealthFlag::FAILED_EDS_HEALTH)
+              : absl::nullopt);
```
Any reason to not do the health flag checks in the function?

alyssawilk commented on source/common/upstream/upstream_impl.cc:
```diff
@@ -645,6 +645,114 @@ ClusterInfoImpl::ResourceManagers::load(const envoy::api::v2::Cluster& config,
+}
+
+void PriorityStateManager::updateClusterPrioritySet(
+    const uint32_t priority, HostVectorSharedPtr current_hosts,
```
&& current_hosts if we're going to take ownership?

dio commented:
@alyssawilk @zuercher thanks for your time for reviewing this code! I think I have addressed your comments. When you have time please have a look. Thank you!

dio commented on source/common/upstream/upstream_impl.h:
```diff
@@ -493,6 +492,56 @@ class ClusterImplBase : public Cluster, protected Logger::Loggable<Logger::Id::u
+                          Network::Address::InstanceConstSharedPtr address,
+                          const envoy::api::v2::endpoint::LocalityLbEndpoints& locality_lb_endpoint,
+                          const envoy::api::v2::endpoint::LbEndpoint& lb_endpoint,
+                          const Upstream::Host::HealthFlag health_checker_flag);
```
Thanks, @zuercher. Added it here: 3d806d2.

dio commented:
@alyssawilk @zuercher Please let me know if there is anything else that I should change. Thanks!

akonradi commented:
Opened #3856 to fix some off-the-end indexing in `PriorityStateManager::updateClusterPrioritySet`.
